Added ConcreteExpansionMicrocrackingDamage class

agiorla opened the pull request:
Added a ConcreteExpansionMicrocrackingDamage class to simulate damage caused by expansion (ASR or RIVE).

closes #43

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/221780/) on 15e5e5f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/blackbear/docs/PRs/46/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/blackbear/docs/PRs/46/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 359a40d7661c2be72f22e1b7babdd3d59c687f97`

moosebuild commented:
Job [Test](https://civet.inl.gov/job/221784/) on 5924d44 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/blackbear/docs/PRs/46/site)

This comment will be updated on new commits.

bwspenc commented on doc/content/source/materials/ConcreteASREigenstrain.md:
```diff
@@ -203,23 +203,28 @@ rotating $\Delta\varepsilon_{i}^{ASR}$ into the current coordinates via
-## Reduction of elastic modulus and tensile strength
+## ASR-induced Damage
+
+The ASR-induced degradation of the mechanical properties of concrete can be simulated with a specific damage models.
```
Reword "a specific damage models".

bwspenc commented on doc/content/source/materials/ConcreteASREigenstrain.md:
```diff
@@ -203,23 +203,28 @@ rotating $\Delta\varepsilon_{i}^{ASR}$ into the current coordinates via
-are the corresponding residual fractional values when the concrete has fully reacted, i.e. $\varepsilon_{ASR}$ tends to $\varepsilon_{ASR}^{\infty}$.  Both $\beta_E$ and $\beta_f$ are parameters input by the user.
 
-Note that the effect of ASR on elasticity is computed in a separate model, as described below. Also, the effect of ASR on the tensile strength is accounted for in the computation of the ASR expansion, but there is not currently a mechanism for having that influence the behavior of damage models. That is a topic of future development.
+In which $\mathbb{E}$ is the fourth-order stiffness tensor of the material and $d^{ASR}$ the ASR-induced damage.
```
'In' -> 'in'

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+
+## Description
+
+The `ConcreteExpansionMicrocrackingDamage` model computes a scalar damage index $d$ that increases with the progress of the eigenstrain defined by the `eigenstrain_name` input parameter (for example, the eigenstrain induced by alkali-silica reaction (ASR) using a [ConcreteASREigenstrain](ConcreteASREigenstrain.md) model) and the current stress state of the material.
```
Close the left parentheses '(for example...'

agiorla replied:
It is closed after 'model) ...'

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+More specifically, this model is aimed at capturing anisotropic effects on ASR (and similar expansion mechanisms) without relying on an explicit redistribution of the eigenstrain itself.
+Instead, external stresses increase the damage so that the \emph{apparent} expansion is negated in the direction of the load.
+However, the eigenstrain still applies pressure in all directions.
+This model uses the following assumptions:
```
'uses' -> 'makes'

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+This model uses the following assumptions:
+
+- The eigenstrain is isotropic.
+- In unconfined condition, the damage increases with the expansion.
```
In unconfined -> In an unconfined

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+
+### Unconfined Damage
+
+In unconfined condition $(\sigma_c = 0)$, the damage $d_{unconfined}$ depends only on the eigenstrain:
```
'unconfined condition' -> 'an unconfined condition'

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+d_{unconfined} = 1.0 - \frac{\varepsilon_{branch}}{\varepsilon_{lin}^{eig} - (\varepsilon_{branch} + \varepsilon_{init})}
+\end{equation}
+
+In which $\varepsilon_{init}$ is the linear expansion at which damage initiates and $\varepsilon_{branch}$ controls the rate at which damage increases with the eigenstrain.
```
'In' -> 'in'

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+
+### Confined Damage
+
+In confined condition (when $\sigma_c > \sigma_u$), the damage is further increased so that the apparent expansion in the direction of the load
```
'In confined' -> 'In a confined'

bwspenc commented on doc/content/source/materials/ConcreteExpansionMicrocrackingDamage.md:
```diff
@@ -0,0 +1,76 @@
+
+This model only computes a scalar damage index, and must rely on an external model to compute the expansion.
+If a [ConcreteASREigenstrain](ConcreteASREigenstrain.md) model (or similar model) is used to simulate ASR, then the `expansion_type` for the ASR model should be set to `Isotropic`.
+In order for the damage to be applied, the stress needs to be computed using a [ComputeDamageStress](ComputeDamageStress.md) stress calculator.
```
'applied' -> 'accounted for in the stress calculation'

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+  params.addClassDescription("Scalar damage model based on extent of internal expansion");
+
+  params.addRequiredParam<MaterialPropertyName>(
+      "eigenstrain_name", "name of the eigenstrain driving the microcracking damage process");
```
The descriptions of all of these parameters should start out with a capitalized first word for consistency.

Also, I have several parameter renaming suggestions that would probably be good to discuss by phone. How about 'microcracking_eigenstrain_name' for this one (to make it clear which eigenstrain is needed, since there could be multiple)?

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+  params.addRequiredParam<MaterialPropertyName>(
+      "eigenstrain_name", "name of the eigenstrain driving the microcracking damage process");
+
+  params.addParam<bool>("use_isotropic_expansion",
```
How about 'assume_isotropic_expansion'?

Also, 'decomposes the expansion alongside the principal direction' -> 'computes linear expansion based on the first principal eigenstrain'

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+                        "indicates whether the model assumes an isotropic expansion (true) or "
+                        "decomposes the expansion alongside the principal direction (false)");
+
+  params.addParam<bool>("use_stress_control",
```
'use_stress_control' -> 'include_confinement_effects'?

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+                        true,
+                        "indicates whether the damage is affected by the current stress state");
+
+  params.addParam<Real>("correction_factor",
```
'correction_factor' -> 'eigenstrain_factor'?

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+                        "evaluating the damage");
+
+  params.addRequiredRangeCheckedParam<Real>(
+      "microcracking_strain_init",
```
'microcracking_initiation_strain'?

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+      "linear strain at which the microcracking initiates (in [m/m])");
+
+  params.addRequiredRangeCheckedParam<Real>(
+      "microcracking_strain_branch",
```
It's not clear to me where the name 'branch' comes from in this parameter.  We should probably either rename this parameter, or somehow explain that.

agiorla replied:
This parameter controls the slope of the damage as a function of the expansion. It has the same dimension as a strain. I wasn't sure how to name it either...

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,167 @@
+      "parameter controlling the rate at which the microcracking increases (in [m/m])");
+
+  params.addRangeCheckedParam<Real>(
+      "expansion_stress_limit",
```
Does this actually need to be greater than some tolerance? It looks like you could get close to a divide by zero condition.

agiorla replied:
This is generally in the order of magnitude of several MPa. We could set it > 1 for safety?

agiorla commented on doc/content/source/materials/ConcreteASREigenstrain.md:
```diff
@@ -203,23 +203,28 @@ rotating $\Delta\varepsilon_{i}^{ASR}$ into the current coordinates via
+  \begin{equation}
+  d^{ASR}(t, \theta) = (1 - \beta_E)\xi(t, \theta)
+  \end{equation}
+- Using the [ConcreteExpansionMicrocrackingDamage](ConcreteExpansionMicrocrackingDamage.md) model, in which the damage is computed as a function of the total ASR strain and the current stress. In this approach, the anisotropic effects of ASR are accounted indirectly via the damage model rather than the eigenstrain itself, and therefore this model should be used with the `expansion_type = Isotropi ...
```
I know i might sound a little bit confusing. The scalar damage that represents the microcracking is affected by the stress state of the material, so that the expansion under load is anisotropic with respect to the loading direction. This can be understood as a "partially anisotropic" model: the expansion is anisotropic but the damage is not. Here, the "expansion" is understood as the apparent expansion under load, which is the total strain at any given time minus the elastic strain at t = 0. Since damage increases with ASR/time, the "expansion" is thus a combination of the true eigenstrain (isotropic) and the additional elastic strain caused by the loss of stiffness (which is anisotropic if the stress tensor is). I'll try to rephrase this.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/228755/) on c813b00 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/blackbear/docs/PRs/46/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/blackbear/docs/PRs/46/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 359a40d7661c2be72f22e1b7babdd3d59c687f97`

agiorla commented:
@bwspenc the test still fails because there is a file missing that is defined in the other PR (so the other PR needs to be merged first). I addressed most of your comments I think, but it is probably a good idea to have a discussion about it.

bwspenc commented on src/materials/ConcreteExpansionMicrocrackingDamage.C:
```diff
@@ -0,0 +1,170 @@
+  params.addRangeCheckedParam<Real>(
+      "expansion_stress_limit",
+      "expansion_stress_limit > 0",
+      "Upper bound compressive stress beyond which damage is controlled by the external stress");
```
The second argument to addRangeCheckedParam() is a default value, but you're missing that here. Since you want this model to work with any unit system, I don't think you want to have a default value. Since you can't have it default to 0 (because it would break the range check), I'd probably switch to addParam(), and then if 'include_confinement_effects' is true, require that this be specified, and in the constructor code enforce that it is greater than zero.

agiorla commented:
@bwspenc I did the rebase and suddenly all these commits from MooseTest have appeared. I am not sure why, or whether that would cause any issue?

bwspenc commented:
Try rebasing again -- that usually makes them go away for me. I'm not really sure why that happens.

agiorla commented:
Just to double-check, from which source should I rebase? is that upstream/devel?

bwspenc commented:
Yeah, do 'git pull --rebase upstream devel', and then something like 'git push -f origin concrete_expansion_microcracking_46'

agiorla commented:
OK, I probably rebased on the upstream master instead of upstream devel. There were some issue with the doc still, I will fix that this afternoon.

agiorla commented:
@bwspenc it seems that it is the doc for the `ConcreteExpansionFunctionEigenstrain` that is causing the test to fail. The file seems located in a weird place, and I'm not sure where it should go?

bwspenc commented:
You'll need to move the file: doc/content/documentation/systems/Materials/ConcreteExpansionFunctionEigenstrain.md to doc/content/source/materials/
The system was recently re-worked to move things around to better match the layout in the source tree.

bwspenc commented on doc/content/source/materials/ConcreteExpansionFunctionEigenstrain.md:
```diff
@@ -0,0 +1,23 @@
+The actual eigenstrain can either be `Isotropic` or `Anisotropic` depending on the `expansion_type`
+parameter. In the latter case, the eigenstrains are redistributed when the material is subject
+to structural loads, following the model developped by [cite:saouma_constitutive_2006] for
+Alkali Silica Reaction. See the documentation for the [`ConcreteASREigenstrain`](/Materials/ConcreteASREigenstrain.md)
```
I think you need to drop the '/Materials/' here for the link to work.

bwspenc commented on doc/content/source/materials/ConcreteExpansionFunctionEigenstrain.md:
```diff
@@ -0,0 +1,23 @@
+The `ConcreteExpansionFunctionEigenstrain` imposes an eigenstrain in which the total volumetric
+strain is computed with a MOOSE [`Function`](/Functions/index.md) object defined by the `function`
+input parameter. The `prefix` parameter is used to distinguish the `volumetric_strain` variable
+defined by other models such as a [`ConcreteASREigenstrain`](/Materials/ConcreteASREigenstrain.md)
```
I think you need to drop the '/Materials/' here for the link to work.

bwspenc commented on doc/content/source/materials/ConcreteExpansionFunctionEigenstrain.md:
```diff
@@ -0,0 +1,23 @@
+# Concrete Expansion Function Eigenstrain
+
+!syntax description /Materials/ConcreteExpansionFunctionEigenstrain
```
The doc build is failing because the source file for this model doesn't exist.

agiorla commented:
@bwspenc I just noticed that this doc file is for a class that is in the blackbear/tests folder. Should I move it somewhere there, or should I just remove it?

bwspenc commented on test/tests/concrete_expansion_microcracking/tests:
```diff
@@ -0,0 +1,88 @@
+                BCs/load_y/value=-10e6
+                BCs/load_z/value=-10e6
+                Outputs/file_base=concrete_expansion_microcracking_10_10_10_out'
+    allow_test_objects = True
```
We have a new system in MOOSE for our software quality that requires that every test provide 3 parameters: a requirement, and issue, and a reference to a design document (the .md file). They're requiring that in all new tests in MOOSE, and we'll start doing that in BlackBear at some point. I'm going to go ahead and merge this, but it will be easiest to get this added to these tests (and others that you have added) while it's fresh on your mind.  I'd really appreciate it if you could do that.  Look in moose/modules/xfem for examples (just because we recently added that stuff to every test in there).
